Event signing and signature validation

reinkrul opened the pull request:
Fixes #25 
Fixes #60

woutslakhorst commented on pkg/db/db.go:
```diff
@@ -74,6 +75,7 @@ type Vendor struct {
 // GetActiveCertificates looks up the vendor's certificates and returns them sorted, longest valid certificate first.
 // Expired certificates aren't returned.
 func (v Vendor) GetActiveCertificates() []*x509.Certificate {
+	// TODO: Copy keyset
```
a todo

reinkrul replied:
Fixed

woutslakhorst commented on docs/pages/architecture/registry.rst:
```diff
@@ -0,0 +1,72 @@
+Nuts registry architecture
+##########################
+
+.. marker-for-readme
```
copy pasted these lines?

reinkrul replied:
You found out my secret...

woutslakhorst commented on docs/pages/architecture/registry.rst:
```diff
@@ -0,0 +1,72 @@
+      IssuedAt (timestamp)
+      Type (string enum)
+      Signature (JWS)
+      Payload (object)
```
How does the 'payload' relate to JWS.payload?

reinkrul commented:
> 2. add an issue which is solved by this PR and add the issue to 0.13.0. So we don't miss anything in the release notes.
Added issues to past PRs (and this one) and I added those to 0.13.
> 3. there now exists 2 types of old entries. Migrating them will have to be done soon, otherwise lots of old stuff and logic..
I intend to do update events next, so then we can migrate old events.

woutslakhorst commented on docs/pages/architecture/registry.rst:
```diff
@@ -0,0 +1,72 @@
+.. _nuts-registry-architecture:
+
+Nuts registry architecture
```
forgot to mention, that is more technical than architectural documentation

woutslakhorst replied:
I still think this document must be under /technical/

reinkrul replied:
Moved it.

woutslakhorst commented on docs/pages/architecture/registry.rst:
```diff
@@ -40,6 +40,8 @@ The signatures are in the JWS (`JSON Web Signature <https://tools.ietf.org/html/
 the JWS will contain an X.509 certificate (in the *x5c* field) describing the entity owning the key. The algorithm
 used for constructing the JWS is RS256 (RSA with SHA-256 hashing function).
 
+The JWS contains also contains the actual payload of the event.
```
dupl words

woutslakhorst commented on docs/pages/architecture/registry.rst:
```diff
@@ -51,6 +53,7 @@ To validate an event signature, the following checks must be performed:
 5. Is the owning entity of the certificate (e.g. a vendor or organization) the one we expected to sign the certificate (see *Owner check* in the table below)?
 6. Is the JWS signed using the RS256 algorithm?
+7. Is the used RSA key of sufficient length (>=2048 bits)?
 7. Is the JWS signed with the private key belonging to the public key in the certificate?
```
Does ReST handle numbering automatically, otherwise this must be an 8?

reinkrul replied:
Fixed.

woutslakhorst commented on go.mod:
```diff
@@ -20,3 +20,5 @@ require (
 	github.com/stretchr/testify v1.4.0
 )
+
+replace github.com/nuts-foundation/nuts-crypto => ../nuts-crypto
```
TODO, wait for other PR's to be merged

reinkrul replied:
Fixed it.

woutslakhorst commented on pkg/cert/util.go:
```diff
@@ -0,0 +1,30 @@
+// GetVendorSubjectAltName tries to get the vendor ID from the certificate SAN.
+// If it doesn't exist, an empty string is returned without error.
+// If an error occurs, an empty string is returned alongside the error.
+func GetVendorSubjectAltName(certificate *x509.Certificate) (string, error) {
```
Missing tests for these 2 funcs

woutslakhorst commented on pkg/events/domain/domain.go:
```diff
@@ -141,3 +175,18 @@ func (t *trustStore) handleEvent(event events.Event) error {
 }
+
+// copyKeys is needed since the jwkSet.extractMap consumes the contents
+// TODO: Temporary function
```
waiting for PR?

reinkrul replied:
Yes, just fixed it.

woutslakhorst commented on pkg/events/domain/domain.go:
```diff
@@ -50,11 +54,24 @@ type RegisterVendorEvent struct {
 	if r.Domain == "" {
 		r.Domain = FallbackDomain
 	}
+	for _, key := range copyKeys(r.Keys) {
```
Shouldn't this validation be skipped if no domain exists? Since that would suggest legacy data for which the SubjectAlternativeName doesn't exist

woutslakhorst commented on pkg/events/domain/domain.go:
```diff
@@ -71,6 +88,22 @@ type VendorClaimEvent struct {
 }
 
+func (v VendorClaimEvent) PostProcessUnmarshal(event events.Event) error {
+	for _, key := range copyKeys(v.OrgKeys) {
```
Is this backwards compatible?
